This notebook contains:

01) Using the groupby function

02) Aggregating values such as mean, median, min and max


01. starting new notebook

In [1]:
# Importing libraries
import pandas as pd
import numpy as np
import os

02. Establishing path and importing dataframe

In [2]:
# Establishing a path to import data frames
path = r'C:\Users\Paola\Desktop\Shaul\Data\CF\August-2023 Instacart Basket Analysis'

In [3]:
path

'C:\\Users\\Paola\\Desktop\\Shaul\\Data\\CF\\August-2023 Instacart Basket Analysis'

In [41]:
# Importing 'orders_products_variablesfixed.pkl' data set
df_ords_prods = pd.read_pickle(os.path.join(path, '2-Data', 'Prepared Data', 'orders_products_variablesfixed.pkl'))

In [42]:
# Checking dimensions and variables of the dataframe
df_ords_prods.shape

(32435070, 16)

In [43]:
df_ords_prods.head(10)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329.0      1.0           1.0                 2.0                8.0   
1  2398795.0      1.0           2.0                 3.0                7.0   
2   473747.0      1.0           3.0                 3.0               12.0   
3  2254736.0      1.0           4.0                 4.0                7.0   
4   431534.0      1.0           5.0                 4.0               15.0   
5  3367565.0      1.0           6.0                 2.0                7.0   
6   550135.0      1.0           7.0                 1.0                9.0   
7  3108588.0      1.0           8.0                 1.0               14.0   
8  2295261.0      1.0           9.0                 1.0               16.0   
9  2550362.0      1.0          10.0                 4.0                8.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                1.0        0.0   
1                    15.0         196                1.0        1.0   
2                    21.0         196                1.0        1.0   
3                    29.0         196                1.0        1.0   
4                    28.0         196                1.0        1.0   
5                    19.0         196                1.0        1.0   
6                    20.0         196                1.0        1.0   
7                    14.0         196                2.0        1.0   
8                     0.0         196                4.0        1.0   
9                    30.0         196                1.0        1.0   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda      77.0            7.0     9.0  Mid-range product   
1         Soda      77.0            7.0     9.0  Mid-range product   
2         Soda      77.0            7.0     9.0  Mid-range product   
3         Soda      77.0            7.0     9.0  Mid-range product   
4         Soda      77.0            7.0     9.0  Mid-range product   
5         Soda      77.0            7.0     9.0  Mid-range product   
6         Soda      77.0            7.0     9.0  Mid-range product   
7         Soda      77.0            7.0     9.0  Mid-range product   
8         Soda      77.0            7.0     9.0  Mid-range product   
9         Soda      77.0            7.0     9.0  Mid-range product   

     Busiest days   Busiest hours  
0  Regularly busy  Average orders  
1    Slowest days  Average orders  
2    Slowest days     Most orders  
3    Slowest days  Average orders  
4    Slowest days     Most orders  
5  Regularly busy  Average orders  
6   Busiests days     Most orders  
7   Busiests days     Most orders  
8   Busiests days     Most orders  
9    Slowest days  Average orders

03. Starting Task 4.8
In this Exercise, you learned how to find the aggregated mean of the “order_number” column grouped by “department_id” for a subset of your dataframe. Now, repeat this process for the entire dataframe.

In [44]:
# Question 2 => perform grouping and aggregation (find the mean)
# Performing a single aggregation
df_ords_prods.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1.0              15.457838
2.0              17.277920
3.0              17.170395
4.0              17.811403
5.0              15.215751
6.0              16.439806
7.0              17.225802
8.0              15.340650
9.0              15.895474
10.0             20.197148
11.0             16.170638
12.0             15.887671
13.0             16.583536
14.0             16.773669
15.0             16.165037
16.0             17.665606
17.0             15.694469
18.0             19.310397
19.0             17.177343
20.0             16.473447
21.0             22.902379

Question 3: Comparing the results of this aggregation with the results of the same step, when runned on a subset of 1000000 rows.
I have noticed that: 1) On the subset, only departments 4,7,13,14,16,17,19 and 20 were present. It is 8 from 21 departments. It indicates that within the first million rows, orders were made only from those departments. It is an interesting insight which is worth looking into. 2) As expected, in the whole dataframe, all 21 departments were present and accounted for. 3) Interestingly, in the subset all means were higher than the same results achieved by the same command executed in the whole dataframe. Most likely, it is the result of the entire data influencing the means. It would be very interesting to run the same command, this time on the median, to see if the results are similar.

Results of the comparison ['department id', 'mean on subset', 'mean on dataframe'].

4	18.8		17.88

7 	17.4		17.22

13	17.9		16.58

14	19.24		16.77

16	19.46		17.66

17	11,29		15.69

19	19.30		17.17

20	17.59		16.47


Question 4
Follow the instructions in the Exercise for creating a loyalty flag for existing customers using the transform() and loc() functions.

In [45]:
# Question 4: Creating a new column, after grouping and transforming the data variable
df_ords_prods['max_order'] = df_ords_prods.groupby(['user_id'])['order_number'].transform(np.max)

In [46]:
# Checking the results
df_ords_prods.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0  2539329.0      1.0           1.0                 2.0                8.0   
1  2398795.0      1.0           2.0                 3.0                7.0   
2   473747.0      1.0           3.0                 3.0               12.0   
3  2254736.0      1.0           4.0                 4.0                7.0   
4   431534.0      1.0           5.0                 4.0               15.0   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                     7.0         196                1.0        0.0   
1                    15.0         196                1.0        1.0   
2                    21.0         196                1.0        1.0   
3                    29.0         196                1.0        1.0   
4                    28.0         196                1.0        1.0   

  product_name  aisle_id  department_id  prices    price_range_loc  \
0         Soda      77.0            7.0     9.0  Mid-range product   
1         Soda      77.0            7.0     9.0  Mid-range product   
2         Soda      77.0            7.0     9.0  Mid-range product   
3         Soda      77.0            7.0     9.0  Mid-range product   
4         Soda      77.0            7.0     9.0  Mid-range product   

     Busiest days   Busiest hours  max_order  
0  Regularly busy  Average orders       10.0  
1    Slowest days  Average orders       10.0  
2    Slowest days     Most orders       10.0  
3    Slowest days  Average orders       10.0  
4    Slowest days     Most orders       10.0

In [47]:
# Creating a flag column based in Customer loyalty
df_ords_prods.loc[df_ords_prods['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [48]:
df_ords_prods.loc[(df_ords_prods['max_order'] <= 40) & (df_ords_prods['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [49]:
df_ords_prods.loc[df_ords_prods['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [50]:
df_ords_prods['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891507
Loyal customer      10294027
New customer         6249525
NaN                       11
Name: loyalty_flag, dtype: int64

I investigated the missing values, they are missing from the following columns: 'product name', 'department id', aisle id', prices' and 'price range loc'. I have tried different ways to drop thos values, since they would not affect the overal analysis. However, I have not been succesful. In order to continue the task, I have left them as it is.

In [14]:
# Checking code's execution
df_ords_prods[['user_id', 'loyalty_flag', 'order_number']].head(15)

user_id      loyalty_flag  order_number
0       1.0      New customer           1.0
1       1.0      New customer           2.0
2       1.0      New customer           3.0
3       1.0      New customer           4.0
4       1.0      New customer           5.0
5       1.0      New customer           6.0
6       1.0      New customer           7.0
7       1.0      New customer           8.0
8       1.0      New customer           9.0
9       1.0      New customer          10.0
10     15.0  Regular customer          15.0
11     15.0  Regular customer          17.0
12     15.0  Regular customer          18.0
13     15.0  Regular customer          21.0
14     15.0  Regular customer          22.0

In [27]:
# Creating a new dataframe without the missing values
df_ords_prods = df_ords_prods[df_ords_prods['price_range_loc'].isnull() == False]

Question 5
The marketing team at Instacart wants to know whether there’s a difference between the spending habits of the three types of customers you identified. Use the loyalty flag you created and check the basic statistics of the product prices for each loyalty category (Loyal Customer, Regular Customer, and New Customer). What you’re trying to determine is whether the prices of products purchased by loyal customers differ from those purchased by regular or new customers.

In [15]:
# Question 5: Check the spending habits of customers based on their loyalty. Find out basic statistics.
df_ords_prods.groupby('loyalty_flag').agg({'prices': ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

In [16]:
# Checking the proportion of customers to find the difference of proportion in the 'average'
df_ords_prods['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891507
Loyal customer      10294027
New customer         6249525
NaN                       11
Name: loyalty_flag, dtype: int64

Comments:
Loyal customers = 32%. Regular customres = 49%, and New customers = 19%. 
In matters of spending, the average prices of products bought by new customers is $13.3 and loyal customers is only $10.4. Regular customers' average product price spending is $12.5 which is in-between.

In [53]:
# This cell (and the next one) are comments made by my mentor Dania. I have added them to add clarity to the analysis.
print(df_ords_prods.groupby('loyalty_flag').agg({'prices': ['mean']}).to_markdown(floatfmt=".1f"))

| loyalty_flag     |   ('prices', 'mean') |
|:-----------------|---------------------:|
| Loyal customer   |                 10.4 |
| New customer     |                 13.3 |
| Regular customer |                 12.5 |


In [54]:
higher_price = (13.3/10.4 -1)*100
print('The new customers buy products {:.1f}% more expensive than loyal customers'.format(higher_price))

The new customers buy products 27.9% more expensive than loyal customers


Question 6
The team now wants to target different types of spenders in their marketing campaigns. This can be achieved by looking at the prices of the items people are buying. Create a spending flag for each user based on the average price across all their orders using the following criteria:
If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”

In [17]:
# Question 6: Creating a new variable, after grouping and transforming the data
df_ords_prods['avg_prices'] = df_ords_prods.groupby(['user_id'])['prices'].transform(np.mean)

In [18]:
# Creating a spending flag column based in Customer loyalty
df_ords_prods.loc[df_ords_prods['avg_prices'] < 10, 'spending_flag'] = 'Low spender'

In [19]:
df_ords_prods.loc[df_ords_prods['avg_prices'] >= 10, 'spending_flag'] = 'High spender'

In [20]:
# Checking the results of the spending flag
df_ords_prods['spending_flag'].value_counts(dropna = False)

Low spender     31800381
High spender      634678
NaN                   11
Name: spending_flag, dtype: int64

Results: High spending customers account for only 0.02% of all Instacart customers. There is much need for a campaign to target customers to increase high spend.

Question 7
In order to send relevant notifications to users within the app (for instance, asking users if they want to buy the same item again), the Instacart team wants you to determine frequent versus non-frequent customers. Create an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column. 

In [21]:
# Question 7: Finding out the median of the 'days since prior order column', grouped by user
df_ords_prods['median_prior_order'] = df_ords_prods.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [22]:
# Creating a flag column based in Customer frequency
df_ords_prods.loc[df_ords_prods['median_prior_order'] > 20, 'frequency_flag'] = 'Non-frequent customer'

In [23]:
df_ords_prods.loc[(df_ords_prods['median_prior_order'] <= 20) & (df_ords_prods['median_prior_order'] > 10), 'frequency_flag'] = 'Regular customer'

In [24]:
df_ords_prods.loc[df_ords_prods['median_prior_order'] <= 10, 'frequency_flag'] = 'Frequent customer'

In [25]:
# Checking the results of the frequency flag
df_ords_prods['frequency_flag'].value_counts(dropna = False)

Frequent customer        22809921
Regular customer          6935132
Non-frequent customer     2690006
NaN                            11
Name: frequency_flag, dtype: int64

Frequent customers account for 70% of all Instacart customers.

08. Exporting new file to Pickle

In [26]:
df_ords_prods.to_pickle(os.path.join(path, '2-Data','Prepared Data', 'orders_products_readyvariables.pkl'))

Data frame exported succesfully. Thank you.